container_opts_unix.go

```diff
@@ -42,6 +56,19 @@ func WithRemappedSnapshotView(id string, i Image, uid, gid uint32) NewContainerO
 	return withRemappedSnapshotBase(id, i, uid, gid, true)
 }
 
+func withLabelRemappedSnapshot(id string, i Image, uid, gid uint32, readonly bool) NewContainerOpts {
+	opts := []snapshots.Opt{
+		snapshots.WithLabels(map[string]string{
+			"containerd.io/snapshot/uidmapping": fmt.Sprintf("%d:%d:%d", 0, uid, 65535),
```

Comment:
How come this is mapping everything after `uid`?
Can we scope this down?
Also just to note, this is specifically a root remap.

Reply:
First, this isn't anything to do with the actual user namespace mapping; that's handled via the OCI spec and related options. Our filesystem "shifting" (to complement the userns mappings provided to the spec) was patterned after a more simplistic monotonic UID/GID shift (as if your `/etc/sub{u,g}id` files had the most default setup found in most distros--a single entry, starting at 100,000--as an example--and mapping the next 64k). You can see `incrementFS` in this same file showing that same simple model implementation.

This new function is to support letting a snapshotter handle the filesystem ownership mapping via non-brute force/`chown` methods (e.g. FUSE overlay support, experimental `shiftfs` support in Ubuntu, etc.), and so we are starting with the same simple idea of providing the base, offset, and a static 64k block (again, in line with defaults on most distros). I think @AkihiroSuda mentioned that future implementations will need to support more interesting use cases (e.g. just like uid maps can be non-contiguous and different sizes than the default 64k). How the actual snapshotter impl. uses the data in the label is really up to the snapshotter.